rs/tests/nested/nns_recovery/common.rs

```diff
@@ -0,0 +1,562 @@
+/// GuestOS image, that Node Providers would use as input to guestos-recovery-upgrader.
+pub const RECOVERY_GUESTOS_IMG_VERSION: &str = "RECOVERY_VERSION";
+
+pub struct SetupConfig {
+    pub impersonate_upstreams: bool,
+}
+
+pub struct TestConfig {}
+
+fn get_host_vm_names(num_hosts: usize) -> Vec<String> {
+    (1..=num_hosts).map(|i| format!("host-{}", i)).collect()
+}
+
+pub fn assign_unassigned_nodes_to_nns(
+    logger: &Logger,
+    topology: &TopologySnapshot,
+) -> TopologySnapshot {
+    info!(logger, "Adding all unassigned nodes to the NNS subnet...");
+    let nns_subnet = topology.root_subnet();
+    let original_node = nns_subnet.nodes().next().unwrap();
+
+    let new_node_ids: Vec<_> = topology.unassigned_nodes().map(|n| n.node_id).collect();
+    block_on(change_subnet_membership(
+        original_node.get_public_url(),
+        nns_subnet.subnet_id,
+        &new_node_ids,
+        &[original_node.node_id],
+    ))
+    .expect("Failed to change subnet membership");
+
+    info!(
+        logger,
+        "Waiting for new nodes to take over the NNS subnet..."
+    );
+    let new_topology = block_on(topology.block_for_newer_registry_version_within_duration(
+        Duration::from_secs(60),
+        Duration::from_secs(2),
+    ))
+    .unwrap();
+
+    let nns_subnet = new_topology.root_subnet();
+    let num_nns_nodes = nns_subnet.nodes().count();
+    assert_eq!(
+        num_nns_nodes,
+        new_node_ids.len(),
+        "NNS subnet should have {} nodes after removing the original node, but found {} nodes",
+        new_node_ids.len(),
+        num_nns_nodes
+    );
+    for node in nns_subnet.nodes() {
+        node.await_status_is_healthy().unwrap();
+    }
+    info!(logger, "Success: New nodes have taken over the NNS subnet");
+
+    // Readiness wait: ensure the NNS subnet is healthy and making progress
+    let nns_node = nns_subnet.nodes().next().unwrap();
+    info!(
+        logger,
+        "Waiting for NNS subnet to become healthy and make progress after membership changes..."
+    );
+    cert_state_makes_progress_with_retries(
+        &nns_node.get_public_url(),
+        nns_node.effective_canister_id(),
+        &logger,
+        Duration::from_secs(300),
+        Duration::from_secs(10),
+    );
+
+    new_topology
+}
+
+pub fn setup(env: TestEnv, cfg: SetupConfig) {
+    assert_version_compatibility();
+
+    if cfg.impersonate_upstreams {
+        impersonate_upstreams::setup_upstreams_uvm(&env);
+    }
+
+    setup_ic_infrastructure(&env, Some(DKG_INTERVAL));
+
+    let host_vm_names = get_host_vm_names(SUBNET_SIZE);
+    let host_vm_names_refs: Vec<&str> = host_vm_names.iter().map(|s| s.as_str()).collect();
+    setup_nested_vm_group(env.clone(), &host_vm_names_refs);
+
+    for vm_name in &host_vm_names {
+        setup_vector_targets_for_vm(&env, vm_name);
+    }
+}
+
+pub fn test(env: TestEnv, _cfg: TestConfig) {
```

Comment:
Is this function actually called , or going to be called, from multiple call-sites? If not, I would suggest inlining it into its only call-site. Ideally shared code is as small as possible.